diesel/src/sqlite/connection/from_sqlite_value.rs

```diff
@@ -0,0 +1,79 @@
+use super::ffi;
+
+/// The sqlite3_value_*type*-functions are only safe for so-called "protected" values.
+///
+/// Per https://www.sqlite.org/c3ref/value.html, values sent to application-defined functions are protected,
+/// making it safe to use those values with FromSqliteValue.
+pub trait FromSqliteValue {
+    /// * `value` must be a valid pointer
+    /// * `value` must be "protected" as defined in the sqlite documentation
+    fn from_sqlite_value(value: *mut ffi::sqlite3_value) -> Self;
+}
+
+impl FromSqliteValue for i32 {
+    fn from_sqlite_value(value: *mut ffi::sqlite3_value) -> Self {
+        unsafe { ffi::sqlite3_value_int(value) }
+    }
+}
+
+impl FromSqliteValue for i64 {
+    fn from_sqlite_value(value: *mut ffi::sqlite3_value) -> Self {
+        unsafe { ffi::sqlite3_value_int64(value) }
+    }
+}
+
+impl FromSqliteValue for f64 {
+    fn from_sqlite_value(value: *mut ffi::sqlite3_value) -> Self {
+        unsafe { ffi::sqlite3_value_double(value) }
+    }
+}
+
+impl FromSqliteValue for String {
+    fn from_sqlite_value(value: *mut ffi::sqlite3_value) -> Self {
+        use std::{slice, str};
+
+        unsafe {
+            // Call sqlite3_value_text(), then _bytes(), as per https://www.sqlite.org/c3ref/column_blob.html
+            let ptr = ffi::sqlite3_value_text(value);
+            let len = ffi::sqlite3_value_bytes(value);
+            assert!(!ptr.is_null());
+
+            // The buffer must be copied immediately (https://www.sqlite.org/c3ref/value_blob.html):
+            // >Please pay particular attention to the fact that the pointer returned from ... sqlite3_value_text(), ...
+            // >can be invalidated by a subsequent call to sqlite3_value_bytes(), sqlite3_value_bytes16(),
+            // >sqlite3_value_text(), or sqlite3_value_text16().
+
+            // This copy makes this abstraction non-zero cost
+
+            let buf = slice::from_raw_parts(ptr, len as usize);
+
+            str::from_utf8(buf)
+                .expect("SQLite guarantees UTF-8 encoding")
+                .to_string()
+        }
+    }
+}
+
+impl FromSqliteValue for Vec<u8> {
+    fn from_sqlite_value(value: *mut ffi::sqlite3_value) -> Self {
+        use std::slice;
+
+        unsafe {
+            // Call sqlite3_value_blob(), then _bytes(), as per https://www.sqlite.org/c3ref/column_blob.html
```

Comment:
Either this comment is wrong, the comment below is wrong, or SQLite's docs are wrong. Both the comment below and SQLite's docs claim that calling `sqlite3_value_bytes` can invalidate the pointer returned by `sqlite3_value_blob`. This is of course identical to the implementation in Diesel today, which means that this definitely isn't happening (or nobody is using binary on SQLite with Diesel... Or we're relying on UB).. Either way, something is lying. I'd like to either fix the comment here, or fix the implementation.

Reply:
Yeah, the SQLite documentation is confusing at best, contradictory at worst.

https://www.sqlite.org/c3ref/column_blob.html says:

>The safest policy is to invoke these routines in one of the following ways:
>
> * sqlite3_column_text() followed by sqlite3_column_bytes()
> * sqlite3_column_blob() followed by sqlite3_column_bytes()
> * sqlite3_column_text16() followed by sqlite3_column_bytes16()
>
>In other words, you should call sqlite3_column_text(), sqlite3_column_blob(), or sqlite3_column_text16() first to force the result into the desired format, then invoke sqlite3_column_bytes() or sqlite3_column_bytes16() to find the size of the result. Do not mix calls to sqlite3_column_text() or sqlite3_column_blob() with calls to sqlite3_column_bytes16(), and do not mix calls to sqlite3_column_text16() with calls to sqlite3_column_bytes().

https://www.sqlite.org/c3ref/value_blob.html says:
>Please pay particular attention to the fact that the pointer returned from sqlite3_value_blob(), sqlite3_value_text(), or sqlite3_value_text16() can be invalidated by a subsequent call to sqlite3_value_bytes(), sqlite3_value_bytes16(), sqlite3_value_text(), or sqlite3_value_text16().

It seems that the documentation is imprecise and that the current implementation is recommended. It looks like the problems would only appear if we cause SQLite to transform a data representation, but they stop short of actually promising that in the documentation.

Perhaps we could add a comment explaining that while the documentation looks contradictory, we trust this implementation because it is clearly recommended by the documentation in https://www.sqlite.org/c3ref/column_blob.html ?